designs/2023-rule-options-defaults/README.md

```diff
@@ -30,144 +30,100 @@ For example:
-- If the config value contains `type: "object"`, each property's `default` will recursively go through this same defaulting logic
-  - If a rule config has not provided a value for the object, a new empty object will be created
-
-Objects created by this defaulting logic should always be shallow cloned (`{ ...original }`) before being passed to rules.
-
-The original `options` provided to rules will be available on a new `context.optionsRaw` property.
-That property will be equivalent to what current versions of ESLint provide as `context.options`.
+Currently, ESLint rule options are passed through a [`getRuleOptions` function](https://github.com/eslint/eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/lib/linter/linter.js#L740) whose only processing is to remove severity.
+That function could be augmented to also take in the `rule.meta.defaultOptions`, if it exists.
 
-### Example: Primitive Value Default
+Defaulting logic could go with the following logic to approximate user intent:
 
-Given a rule schema that defines a default value:
+1. If the user-provided value is `undefined`, go with the default option
+2. If the user-provided value is an object, a new `{}` object is created with values recursively merged
+3. Arrays and non-`undefined` literals -including `null`- are used as-is
 
-```json5
-{
-  default: "implicit",
-  type: "string",
-}
-```
+See [[Reference] feat: add meta.defaultOptions](https://github.com/eslint/eslint/pull/17656) > [`function deepMerge`](https://github.com/eslint/eslint/pull/17656/files#diff-9bbcc1ca9625b99554a055ac028190e5bd28432207a7f1a519690c5d0484287fR24) and [`describe("deepMerge")`](https://github.com/eslint/eslint/pull/17656/files#diff-fa70d7d4c142a6c3e727cf7280f981cb80a1db10bd6cc85c205e47e47da05cfeR25) for the proposed defaulting logic.
 
-The proposed change would impact only cases when the provided value is `undefined`:
-
-| Rule Config             | Current `context.options` | Proposed `context.options` |
-| ----------------------- | ------------------------- | -------------------------- |
-| `"error"`               | `[]`                      | `["implicit"]`             |
-| `["error"]`             | `[]`                      | `["implicit"]`             |
-| `["error", "explicit"]` | `["explicit"]`            | `["explicit"]`             |
-
-### Example: Nested Object Defaults
-
-Using [`class-methods-use-this`](https://github.com/eslint/eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/lib/rules/class-methods-use-this.js)'s schema as an example:
-
-```json5
-[
-  {
-    type: "object",
-    properties: {
-      exceptMethods: {
-        type: "array",
-        items: {
-          type: "string",
-        },
-      },
-      enforceForClassFields: {
-        type: "boolean",
-        default: true,
-      },
-    },
-    additionalProperties: false,
-  },
-]
-```
+The original `options` provided to rules will temporarily be available on a new `context.optionsRaw` property.
```

Comment:
I'm not sure what the value of `context.optionsRaw` is. If a rule can't adhere to what `meta.defaultOptions` does, wouldn't they just leave it off and continue handling their options as they currently do?

Overall, I'm not a fan of adding a property just to remove it later. What can we do to streamline?

Reply:
My thought was that some rules would want to opt into the benefits of `meta.defaultOptions` before they're able to use the actual defaulted options. As in, rules that would likely need a subtle options breaking change to use `meta.defaultOptions` without `context.optionsRaw`:

* https://github.com/eslint/rfcs/pull/113/files#diff-10a9af2250174111ad17b149de4bcf21278ee271a25a85a3de41e124b6cddba3R70 calls out core ESLint rules (interestingly, all eslint.style candidates)
* https://github.com/eslint/rfcs/pull/113#discussion_r1331458798 mentions two typescript-eslint rules:
    * [`@typescript-eslint/ban-types`](https://typescript-eslint.io/rules/ban-types): `extendDefaults` is used to toggle the default rules. 
        * We _could_ potentially breaking change it to remove that option, and instead tell folks to manually set the built-ins they don't want to `null` instead. As of the last 1-2 years of TypeScript type system improvements, there've been fewer and fewer reasons to not use the rule's defaults...
        * [This Sourcegraph query for `extendsDefault.{1,7}false` in `.*eslint.*` JS/TS files](https://sourcegraph.com/search?q=context%3Aglobal+%2FextendDefaults.%7B1%2C7%7Dfalse%2F+file%3A.*eslint.*+%28lang%3AJavaScript+OR+lang%3ATypeScript%29&patternType=standard&sm=1&groupBy=repo) shows 337 at time of comment; [this same Sourcegraph query for `extendsDefault.{1,7}false` in `.*eslint.*` JS/TS files excluding Microsoft repos (which tend to have many duplicates of older TSLint/typescript-eslint settings)](https://sourcegraph.com/search?q=context:global+/extendDefaults.%7B1%2C7%7Dfalse/+file:.*eslint.*+%28lang:JavaScript+OR+lang:TypeScript%29+-repo:%5Egithub%5C.com/%28microsoft%7Cpnp%7CSharePoint%29&patternType=standard&sm=1&groupBy=repo) shows only 130 results at time of comment
        * In summary:  ✅ I don't think this is a major blocker.
    * [`@typescript-eslint/naming-convention`](https://typescript-eslint.io/rules/naming-convention): this rule has a very convoluted set of options and wide range of impact. Anecdotally, I believe most users who choose to configure it end up overriding quite a lot. ✅ I don't think this is a major blocker.


Overall: I like `context.optionsRaw` as a nicety, but 👍 agreed it doesn't add very much value. I can remove it from the RFC and add a mention of why not to use it.

Reply:
Temporarily adding `context.optionsRaw` would cause two breaking changes which wouldn't be ideal. One because any third-party rules that switches to it would require the user to be on an ESLint version that supports it. And then another when `context.optionsRaw` is removed...

Reply:
We already resolved to remove it. :)